### PR TITLE
<CC_OS> Task defination re-structuring

### PR DESCRIPTION
**Pull request by pranjalchanda08**
Re-structuring of CC_OS TASK definition done.

### CODE REVIEW

File: projects/demo_cc_os.src/project.c [hunk 1]
```diff
@@ -15,10 +15,11 @@
...
 
-/* Define the Tasks */
-CC_TASK_DEF( TASK_A, task_handler, NULL, 10, 100);
-CC_TASK_DEF( TASK_B, task_handler, NULL, 10, 100);
-CC_TASK_DEF( TASK_C, task_handler, NULL, 10, 100);
+extern void task_handler(cc_os_args args);
+
+cc_os_task_t * Task_A;
+cc_os_task_t * Task_B;
+cc_os_task_t * Task_C;
```
Comment (akashkollipara) [inline projects/demo_cc_os.src/project.c:22]: static cc_os... ?

File: src/include/visor/terravisor/cc_os/cc_os_config.mk [hunk 2]
```diff
@@ -11,11 +11,11 @@
...
-CC_OS_TASK_STACK_LEN 	?= 255
+CC_OS_IDLE_TASK_STACK_LEN 	?= 255
 CC_OS_POWER_SAVE_EN 		?= 0
 
 $(eval $(call add_define,CC_OS_USE_DYNAMIC))
 $(eval $(call add_define,CC_OS_HEAP_SIZE))
 $(eval $(call add_define,CC_OS_MAX_THREAD))
-$(eval $(call add_define,CC_OS_TASK_STACK_LEN))
+$(eval $(call add_define,CC_OS_IDLE_TASK_STACK_LEN))
 $(eval $(call add_define,CC_OS_POWER_SAVE_EN))
```
Comment (akashkollipara) [inline src/include/visor/terravisor/cc_os/cc_os_config.mk:21]: Move this file to `src/visor/terravisor/cc_os/config.mk`

File: src/include/visor/terravisor/cc_os/utils/cc_os_sched.h [hunk 3]
```diff
@@ -62,6 +64,8 @@ struct cc_sched_tcb
 	c_char  * name;						///> Name of the Current Task
 	uint8_t   priority;					///> Priority of the task
 	uintptr_t stack_ptr;					///> Stack Pointer
+	task_fn   task_func;					///> Task Call Function
+	uintptr_t args_ptr;					///> Task Call argument ptr
```
Comment (akashkollipara) [inline src/include/visor/terravisor/cc_os/utils/cc_os_sched.h:68]: Just a questions... why don't we make it std that we dont pass to callbacks/tasks/thread, etc...
instead we can write a function to retrieve params.

File: src/visor/terravisor/services/kernel/cc_os_tasks.c [hunk 4]
```diff
@@ -47,13 +47,12 @@ extern cc_sched_tcb_t g_cc_os_tcb_list[];
...
 #endif
 
-CC_TASK_DEF(
-	cc_os_idle,				/* Name of the instance */
-	_cc_os_idle_task_fn,			/* Function pointer */
-	&g_sched_ctrl,				/* Task Args*/
-	CC_OS_IDLE_TASK_PRIORITY,	/* Task Priority */
-	CC_OS_TASK_STACK_LEN		/* Stack Length of IDLE Task */
-);
+#if CC_OS_DYNAMIC == CC_OS_FALSE
```
Comment (akashkollipara) [inline src/visor/terravisor/services/kernel/cc_os_tasks.c:50]: Use normal FALSE or TRUE

**Review by akashkollipara — CHANGES_REQUESTED** [review]

File: src/include/visor/terravisor/cc_os/utils/cc_os_sched.h (hunk 3 above)
Comment (pranjalchanda08) [reply src/include/visor/terravisor/cc_os/utils/cc_os_sched.h:68]: See if we do that them it is very different with how the pthread or any other standard calls are.

Comment (pranjalchanda08) [reply src/include/visor/terravisor/cc_os/utils/cc_os_sched.h:68]: Resolving for now with an open idea. Lets brainstorm over this.

**Review by pranjalchanda08 — COMMENTED** [review]
Resolved Comments

**Comment by pranjalchanda08** [comment]
Closes : #201

File: src/visor/terravisor/services/kernel/cc_os_sched.c [hunk 5]
```diff
@@ -193,18 +193,18 @@ static void __cc_sched_context_switch(cc_sched_tcb_t * next_task)
 static void __cc_sched_wait_list_adjustment(cc_sched_ctrl_t * sched_ctrl)
 {
 	cc_sched_tcb_t * ptr = sched_ctrl->wait_list_head;
-	const size_t * wait_res = (size_t *)ptr->wait_res.wait_on_resource;
+	const long int * wait_res = (long int *)ptr->wait_res.wait_on_resource;
```
[the remark this answers is not in the document]
Comment (akashkollipara) [reply src/visor/terravisor/services/kernel/cc_os_sched.c]: `int` should be fine!

File: src/visor/terravisor/services/kernel/cc_os_sched.c [hunk 6]
```diff
@@ -106,7 +106,7 @@ status_t _insert_before(cc_sched_tcb_t ** ptr, cc_sched_tcb_t * new_node, uint8_
 	CC_OS_ASSERT_IF_FALSE(new_node != CC_OS_NULL_PTR);
 	switch (link_type)
```
Comment (akashkollipara) [inline src/visor/terravisor/services/kernel/cc_os_sched.c:107]: why not if-elseif-else...
switch-case makes a table in rodata section. it's expensive for ram

File: src/visor/terravisor/services/kernel/cc_os_sem.c [hunk 7]
```diff
@@ -31,34 +31,28 @@ extern cc_sched_ctrl_t g_sched_ctrl;
...
 	{
 		return error_memory_low;
 	}
-	else
-	{
-		(*sem_ptr)->sem_val = init_val;
-		(*sem_ptr)->sem_init = CC_OS_TRUE;
-	}
-
 #endif
```
[the remark this answers is not in the document]
Comment (akashkollipara) [reply src/visor/terravisor/services/kernel/cc_os_sem.c:43]: same goes for rest.

[the remark this answers is not in the document]
Comment (pranjalchanda08) [reply src/visor/terravisor/services/kernel/cc_os_sem.c:43]: Let us discuss this. And keep it an open item for now.

**Pull request merged**